### DISCUSSION TITLE
Using sprite sheets to avoid OS related font differences during testing

### DISCUSSION THREAD

**Discussion by matthew16550** [plantuml/plantuml#681] [Ideas]
PlantUML draws fonts slightly differently on Linux, on macOS and on Windows (even when the font files are identical). Text usually has different dimensions and different pixel renderings. It is caused by Java & OS internals that we cannot control.

This makes testing more complicated because the font differences are usually not important to us, but they make the output data different on every OS.

If we can make the test outputs pixel identical for all OS then dimensions calculated from text sizes will always be the same and a single set of reference images can be shared by all developers for regression testing.

As an idea to achieve this I propose we make a new pragma `!pragma testing-font` which switches text rendering to use sprite sheets instead of the `java.awt` text drawing methods:

* Commit to source control some PNG files containing font sprite sheets - grids of characters in a few sizes and plain / italic / bold combinations.

* To keep the PNG files small we use a limited set of characters, I think the basic ASCII set (0x21 - 0x7E) would be good enough for testing.  Any other chars would be rendered as the [tofu](https://english.stackexchange.com/questions/62524/what-do-you-call-the-phenomenon-where-a-rectangle-is-shown-because-a-font-lack) rectangle.  It would be simple to add more characters if needed.

* Text rendering is done by copying characters from a sprite sheet to the target.

* For simplicity, any font names in skinparam / styles will be ignored and all text will be monospaced.

* We select a few font sizes for the sprite sheets.  During `UGraphic.draw()` / `calculateDimension()` etc we could use the nearest sized sprite sheet instead of the actual size in skinparam / styles, I think it will be good enough for most testing.  After looking through the themes & skins I'm initially thinking these sizes: 9, 11, 14, 20.

* Text will not look beautiful, but I think we can make it look okayish.

* Font colors can be supported I suspect without too much effort.

* Decorations like underline & strike should behave the same without any work.

* Rendering this way is probably a lot slower, but it's only for testing so hopefully will not matter too much.

* This would also shield our tests from any text rendering changes in different Java versions.

* We will need to be careful about the licensing of the font we choose for the sprite sheets, I'm initially thinking [JetBrains Mono](https://www.jetbrains.com/lp/mono/) which looks nice and has very clear license terms. (And because I like IntelliJ!)

@arnaudroques what do you think?

**Comment by arnaudroques** [comment]
> @arnaudroques what do you think?
>
It's a great idea!


> If we can make the test outputs pixel identical for all OS then dimensions calculated from text sizes will always be the same and a single set of reference images can be shared by all developers for regression testing.
> 
However, I suspect that, speaking about PNG, there might be other differences between OS. For example, I don't know if antialiasing is managed the same way between Linux and Windows.
So having exactly the same image across all Java & OS version may be very difficult.
BTW, that's why there is now a `UGraphicDebug` class that generate exactly the same output across all OS/Version. However, this output is not user friendly :-)

> As an idea to achieve this I propose we make a new pragma `!pragma testing-font` which switches text rendering to use sprite sheets instead of the `java.awt` text drawing methods:
> 
The more logical would be to add a new `PNG_DEBUG` in `FileFormat` (I try to decouple diagram definition and image generation).
Having said that, this new pragma `!pragma testing-font` could also be implemented to force use of `PNG_DEBUG` output even if `PNG` is specified.


>     * Commit to source control some PNG files containing font sprite sheets - grids of characters in a few sizes and plain / italic / bold combinations.
> 
Maybe italic / bold characters could be generated from plain characters. 

>     * We select a few font sizes for the sprite sheets.  During `UGraphic.draw()` / `calculateDimension()` etc we could use the nearest sized sprite sheet instead of the actual size in skinparam / styles, I think it will be good enough for most testing.  After looking through the themes & skins I'm initially thinking these sizes: 9, 11, 14, 20.
>
Or we could generated the missing size using interpolation.

>     * This would also shield our tests from any text rendering changes in different Java versions.
> 
Yes, but as I said, I am wondering if there will not be other rendering changes.


My advice would be to implement a first proof of concept, very simple (for example, with tofu rectangle for *all* characters), to check if pixel rendering is really the same across Java & OS version.
If you go ahead, I would suggest to create a new package for this new `UGraphic` implementation.

**Comment by matthew16550** [comment]
# Anti-aliased Line & Curve Rendering

You are right, some JREs produce slightly different output when drawing anti-aliased lines & curves. But I think it will not be a significant problem for us.

I have used GitHub matrix actions to [compare](https://github.com/matthew16550/plantuml/actions/runs/1274501900) JRE versions & Operating Systems when using `Graphics2D` to draw this test image:

![raster-engine-reference](https://user-images.githubusercontent.com/39400458/134799330-f952358d-b656-4863-9dcc-cb02318ba969.png)

It is a collection of ovals and lines with various stroke widths / caps / joins both with and without anti-aliasing. Hopefully it contains enough to identify any line & curve differences.  It has certainly found some.

## JRE

These JREs produced pixel identical images:
* Temurin 11.0.12+7
* Temurin 17+35
* Zulu 8.56.0.21-CA
* Zulu 11.50+19-CA
* Zulu 12.3+11-CA
* Zulu 13.42+17-CA
* Zulu 14.29+23-CA
* Zulu 15.34+17-CA
* Zulu 16.32+15-CA
* Zulu 17.28+13-CA

And these were not identical:
* Temurin 8.0.302+8
* Zulu 9.0.7.1
* Zulu 10.3+5

## Operating System

These [OS](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) were tested using all JREs:

* macos-10.15
* macos-11
* ubuntu-18.04
* ubuntu-20.04
* windows-2016
* windows-2019
* windows-2022

For this experiment the OS did not cause any difference.

We have seen the OS cause font differences, but fonts were not part of the experiment and I will focus on them separately.

## Rendering Library

Anti-aliased rendering is done by a pluggable library, there is some history [here](https://github.com/bourgesl/marlin-renderer/wiki/History-and-context).

[Marlin](https://github.com/bourgesl/marlin-renderer) is the current library used by OpenJDK.

Running Java with `-Dsun.java2d.renderer.log=true` will log Marlin version details. If you are not using Marlin then it might not log anything. 

It is possible to [use](https://github.com/bourgesl/marlin-renderer/wiki/How-to-Use) a library that was downloaded separately from the JRE.  I have successfully run Java with several Marlin JARs but have not found a version that exactly matches the common output we are seeing above (there is always a couple of curve pixels slightly different). OpenJDK has its [own copy](https://github.com/openjdk/jdk/tree/master/src/java.desktop/share/classes/sun/java2d/marlin) of the Marlin source, and I suspect it must be slightly different from the Marlin repo where the JARs come from. 

# Discussion

Rendering of lines & curves has been stable for recent OpenJDK releases and does not seem related to OS.

Visual regression testing needs identical output on developer machines and in the CI workflow.  It does not matter for this testing if some JREs create different output provided that developers & Github Actions have reasonable access to JREs that create identical output.

Currently, if we use any recent OpenJDK Temurin or Zulu release on any OS then lines & curves will be ok. Perhaps other JREs will be suitable as well.

In future perhaps OpenJDK will tweak their anti-aliasing algorithm.  If that happens we might be able to ignore it for a long time by staying with an older JRE for testing.  Eventually we would need to update reference images to match the new version but with any luck it would not happen very often.

I am still experimenting with font differences.

**Comment by arnaudroques** [comment]
Thanks for your tries !

Two thoughts:

1. Rather than focusing on PNG, we could use SVG.
2. We could also write our own library that would use directly `BufferedImage` to generate PNG.

Point 2 is not that hard to implement and could be an interesting project. (We only use lines, ellipses, rectangles, cubic lines.)
Image generation would probably by slow and images would not be very nice.
But this way, we could guarantee the same behavior everywhere.

**Reply by matthew16550** [reply]
I don't quite understand, could you add some more detail?

**Comment by arnaudroques** [comment]
Sure!

My 2 points are completely unrelated.

Point 1:
Since the goal is to do non-regression tests and since PNG generation vary from one config to another, we could use SVG image to check non-regression. SVG source will not vary depending on the OS and the JVM (well, except for font.). So using SVG output instead of PNG output will simplify non-regression tests.
And we could use your idea of sprite sheet for font with SVG output.

Point 2:
If you still want to use PNG, we could write our own library (so without using `java.awt`) to create PNG image (it's possible thanks to `BufferedImage` class). This way, PNG rendering would be stable over versions.

Tell me if you need more details.

**Comment by matthew16550** [comment]
I'm focusing on PNG output because in other projects, "Visual Approval Testing" has made my "Code > Test > Code" loop much faster.  

Whenever a test fails, the framework will automatically write the failed image to disk, and straight away we can diff it against a reference image.  IntelliJ has image diffing built-in so I don't leave the IDE.  I think you are using Eclipse, don't remember if that has an image differ?

Text formats like `DEBUG` & `SVG` are great for small diffs but for bigger diffs all they tell me is that lots of tiny details are different.  I won't understand until I see a picture.

This is an example for one style of approval testing, it's real code that works on my branch:

``` java
@ExtendWith(ApprovalTestingJUnitExtension.class)
class Example {

	// ApprovalTestingJUnitExtension will inject this before the test method runs
	private ApprovalTesting approvalTesting;

	@Test
	void test_example() throws Exception {
		final BufferedImage image = plantuml2image(
				"@startuml",
				"a -> b",
				"@enduml"
		);
		approvalTesting.approve(image);
	}
}
```

The first time the test runs it will save a file called `Example.test_example.approved.png`.  We add this to git.

Next time the test runs it will compare `image` against the "approved" file.  If they are different it will save a file called `Example.test_example.failed.png` and the test fails.  Then we compare the images ourselves and adjust the code or if we decide the new image is correct then we rename `... failed.png` to `... approved.png` and commit it to git.

The GitHub workflow is configured to store any "failed" files as an artifact, so we can download them for local comparison.

From my experiment [above](https://github.com/plantuml/plantuml/discussions/681#discussioncomment-1385683) it looks like most recent JREs will render lines & curves identically.  If we do find differences affecting us then yes it might help to write our own rendering code though we might be able to work around it by [using](https://github.com/bourgesl/marlin-renderer/wiki/How-to-Use) a specific version of the Marlin library.

I have basic font sprites working now, the next step is connecting them to `UGraphic` and `StringBounder`.  More details in a separate post.

**Comment by matthew16550** [comment]
An update on the fonts work.

This image was created from 9pt and 20pt sprites:
 
![FontSpriteSheetTest test_stored_font_sprites_on_white failed](https://user-images.githubusercontent.com/39400458/135655211-61f55f40-0327-412e-874c-458e117035ef.png)

The api is currently `FontSpriteSheet.drawString(Graphics g, String s, int x, int y)`.

A lot of tests have run in GitHub looking for differences in JRE / OS.  Currently text from sprites is identical everywhere for every variation of colour & alpha that I can think of, test cases are [here](https://github.com/matthew16550/plantuml/blob/font-sprite-experiment/test/net/sourceforge/plantuml/ugraphic/fontspritesheet/FontSpriteSheetTest.java).  It uses simple blitting (currently no scaling) and from what I have seen of OpenJDK source, that particular code is common across all OS.

There is an occasional pixel colour off-by-one difference when comparing a raw font with a sprite but only with alphas 128...254. It does not affect drawing with the sprites, only relevant when testing that the sprite sheet was created correctly.

**Comment by matthew16550** [comment]
Font sprites are now working with `DriverTextG2d` / `StringBounder`, and I have run lots of tests showing identical output on all OS, so I think it makes sense to progress to a PR?

I haven't looked at missing size interpolation, that could be a later task.

The big outstanding question is how to enable it.  Currently, it uses a static variable.  I suggested a `!pragma` because consistent string bounds will be helpful for testing many file formats, it's a universal thing kind of like `elk` / `smetana`.  Perhaps the only place where the sprite images will help is `PNG` but the `StringBounder` could be used everywhere.

@arnaudroques ?

**Reply by arnaudroques** [reply]
>  I suggested a `!pragma` because consistent string bounds will be helpful for testing many file formats,

I agree. `!pragma` is the right solution.

**Comment by matthew16550** [comment]
@arnaudroques 

(Getting close to some PRs now)

If we add the font-sprite PNGs to `plantuml.jar` then it will grow by 94Kbytes.  That's for all combinations of plain / italic / bold / bold-italic, with 9 / 11 / 14 / 20 point sizes.

Shall we do that so `!pragma testing_font` will work for everyone or perhaps put them in a `testResources` dir, so it will only work for test code?

If the latter, then what will we do when `!pragma testing_font` is used and no font-sprite is available, easiest is just to throw an exception?

**Reply by arnaudroques** [reply]
> Shall we do that so `!pragma testing_font` will work for everyone or perhaps put them in a `testResources` dir, so it will only work for test code?

Let's keep it simple and add font-sprite PNG to `plantuml.jar`. Maybe we will be able to compress those files later



> (Getting close to some PRs now)

I'm curious to see the result :-)

**Comment by matthew16550** [comment]
@arnaudroques I was thinking to get back and finish this but not sure how it would fit in with the Haxe rewrite, perhaps it would not be useful any more?

**Reply by arnaudroques** [reply]
About Haxe port, we will target on SVG output first (because it is somehow more portable).
I'm not sure if the Haxe rewrite will be successful, and this will probably take several years.
So I think that your idea is still interesting and will likely be useful when we will target PNG in Haxe.

However, I understand if you prefer to pause this for a few weeks/months, just to wait for the situation to be more clear :-)

**Reply by matthew16550** [reply]
Yes I'll wait to see in a few months.

PlantUML-Haxe might benefit more from this idea than PlantUML-Java.
